### DISCUSSION TITLE
SetOption114 in Mode2 with mcp23017

### DISCUSSION THREAD

**Discussion by hackepeterOli** [arendst/Tasmota#19842] [Support request] [answered]
In my smart home project I use the WT32-ETH01 with two mcp23017 with the new Mode2 driver. 
The two mcp23017 are properly recognised and I have configured 24 relays and 8 buttons:

Rule3 on file#mcp23x.dat do {"NAME":"WT32-ETH01","GPIO":[256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,32,33,34,35,36,37,38,39]} endon

The 8 buttons should only send MQTT messages. So I tested SetOption114 and SwitchMode1 15. Unfortunately, the relay is still controlled with Button1. 

Have I overlooked something?

<img width="395" alt="image" src="https://github.com/arendst/Tasmota/assets/44376749/daf0d5c6-4244-497f-bbce-6d67b40b690c">

Tasmota 13.2.0.1

**Comment by barbudor** [comment]
You have defined the MCP pins as Buttons but you are using commands that are relevant to Switches 

You may have better luck with the proper command [`SetOption73 1`](https://tasmota.github.io/docs/Commands/#setoption73) which is also referenced in the [Button](https://tasmota.github.io/docs/Buttons-and-Switches/#button) section of the docs

**Reply by hackepeterOli** [reply]
Sorry, no luck with SetOption73 1
<img width="464" alt="image" src="https://github.com/arendst/Tasmota/assets/44376749/0c91632f-a9f2-469a-8bc1-b86b69d1c3b6">

**Reply by barbudor** [reply]
You definitively have the Button action sending the MQTT message as expected
But something seems to send that back to the POWER. It's not supposed to happen.

Have you played with ButtonTopic ?
Can you enable WebLog 3 and redo the same ?

And please do not use screenshot but copy the logs using triple-backtics like:
![image](https://github.com/arendst/Tasmota/assets/2996042/66988037-b911-4852-8566-4c0d79a203ae)

**Reply by hackepeterOli** [reply]
Hi,

No luck with ButtonTopic 1 :(
I see the MQTT Messages (Button1_Action, Button2_Action,...) but tasmota switches the POWER also.

here is the log with loglevel 3:
``` 
15:29:58.389 CMD: ButtonTopic 1
15:29:58.391 SRC: WebConsole from 192.168.178.197
15:29:58.392 CMD: Grp 0, Cmd 'BUTTONTOPIC', Idx 1, Len 1, Pld 1, Data '1'
15:29:58.394 CFG: CR 428/699, Busy 0
15:29:58.397 MQT: stat/tasmota_94B2B8/RESULT = {"ButtonTopic":"tasmota_94B2B8"}
15:30:40.870 CMD: SetOption73 1
15:30:40.872 SRC: WebConsole from 192.168.178.197
15:30:40.873 CMD: Grp 0, Cmd 'SETOPTION', Idx 73, Len 1, Pld 1, Data '1'
15:30:40.876 MQT: stat/tasmota_94B2B8/RESULT = {"SetOption73":"ON"}
15:30:41.276 CFG: Saved, Count 35, Bytes 4096
15:30:52.421 MQT: cmnd/tasmota_94B2B8/POWER2 = ON
15:30:52.425 MQT: stat/tasmota_94B2B8/RESULT = {"Button2":{"Action":"SINGLE"}}
15:30:52.530 SRC: MQTT
15:30:52.533 CMD: Grp 0, Cmd 'POWER', Idx 2, Len 2, Pld 1, Data 'ON'
15:30:52.549 MQT: stat/tasmota_94B2B8/RESULT = {"POWER2":"ON"}
15:30:52.551 MQT: stat/tasmota_94B2B8/POWER2 = ON
15:30:53.223 CFG: Saved, Count 36, Bytes 4096
15:30:58.704 MQT: cmnd/tasmota_94B2B8/POWER1 = ON
15:30:58.710 MQT: stat/tasmota_94B2B8/RESULT = {"Button1":{"Action":"SINGLE"}}
15:30:58.726 SRC: MQTT
15:30:58.727 CMD: Grp 0, Cmd 'POWER', Idx 1, Len 2, Pld 1, Data 'ON'
15:30:58.743 MQT: stat/tasmota_94B2B8/RESULT = {"POWER1":"ON"}
15:30:58.745 MQT: stat/tasmota_94B2B8/POWER1 = ON
15:30:59.291 CFG: Saved, Count 37, Bytes 4096
15:31:08.534 MQT: cmnd/tasmota_94B2B8/POWER1 = ON
15:31:08.540 MQT: stat/tasmota_94B2B8/RESULT = {"Button1":{"Action":"CLEAR"}}
15:31:08.555 SRC: MQTT
15:31:08.557 CMD: Grp 0, Cmd 'POWER', Idx 1, Len 2, Pld 1, Data 'ON'
15:31:08.572 MQT: stat/tasmota_94B2B8/RESULT = {"POWER1":"ON"}
15:31:08.575 MQT: stat/tasmota_94B2B8/POWER1 = ON
15:31:17.320 MQT: cmnd/tasmota_94B2B8/POWER1 = ON
15:31:17.327 MQT: stat/tasmota_94B2B8/RESULT = {"Button1":{"Action":"SINGLE"}}
15:31:17.342 SRC: MQTT
15:31:17.344 CMD: Grp 0, Cmd 'POWER', Idx 1, Len 2, Pld 1, Data 'ON'
15:31:17.359 MQT: stat/tasmota_94B2B8/RESULT = {"POWER1":"ON"}
15:31:17.362 MQT: stat/tasmota_94B2B8/POWER1 = ON
```

**Reply by sfromis** [reply]
Seeing `SRC: MQTT` is a clear indicator of Tasmota receiving an incoming command on a MQTT subscription, which might be from Tasmota itself, as `MQT: cmnd/tasmota_94B2B8/POWER1 = ON` indicates that it is publishing to the command topic.

Try turning off that using the command `ButtonTopic 0`. Using `ButtonTopic 1` is generally a bad idea, setting a topic can make sense if you want it to send a command to *another* Tasmota device.

**Reply by barbudor** [reply, marked as answer]
The aim of my question "have you played with Button Topic" was not to encourage you to do so but I was wondering if you had which could have screwed you.

Please restart from a clean setup. I don't see any good reason for MCP's Buttons not to support SetOption73 correctly as native ones.

Are you sure not having any kind of automation that is sending a Power ON through MQTT?

**Reply by sfromis** [reply]
That Power ON command is probably the result of playing with ButtonTopic, as it is also published by Tasmota. Hence my suggestion of clearing it.

**Reply by hackepeterOli** [reply]
Yes, now with ButtonTopic 0 it works. The test above was a fresh system. So the only option i need is SetOption73 1.
Thanks so much!
